pkg/csi/service/vanilla/controller.go

```diff
@@ -906,6 +921,464 @@ func (c *controller) getAccessibleTopologiesForDatastore(ctx context.Context, vc
+					topologySegmentsList, vcHost, err)
+			}
+			if len(sharedDatastores) == 0 {
+				log.Warnf("no shared datastores found for accessibility requirements pertaining to vCenter %q",
+					vcHost)
+				continue
+			}
+			// Filter datastores based on user access.
+			if isAuthCheckFSSEnabled {
+				sharedDatastores, err = c.filterDatastores(ctx, sharedDatastores, vcHost)
+				if err != nil {
+					return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+						"failed to filter datastores based on authorisation check in vCenter %q. Error: %+v",
+						vcHost, err)
+				}
+			}
+			// Call CreateVolume.
+			volumeMgr, err = common.GetVolumeManagerFromVCHost(ctx, c.managers, vcHost)
+			if err != nil {
+				return nil, csifault.CSIInternalFault, logger.LogNewErrorCode(log, codes.Internal, err.Error())
+			}
+			volumeInfo, faultType, err = common.CreateBlockVolumeUtilForMultiVC(ctx,
+				common.VanillaCreateBlockVolParamsForMultiVC{
+					Vcenter:                   vcenter,
+					VolumeManager:             volumeMgr,
+					CNSConfig:                 c.managers.CnsConfig,
+					StoragePolicyID:           storagePolicyID,
+					Spec:                      &createVolumeSpec,
+					SharedDatastores:          sharedDatastores,
+					SnapshotDatastoreURL:      snapshotDatastoreURL,
+					ClusterFlavor:             cnstypes.CnsClusterFlavorVanilla,
+					FilterSuspendedDatastores: filterSuspendedDatastoresEnabled,
+				})
+			if err != nil {
+				log.Errorf("failed to create volume. Error: %+v", err)
+				continue
+			}
+			log.Infof("volume %q created in vCenter %q. Proceeding to calculate "+
+				"accessible topology for the volume", volumeInfo.VolumeID.Id, vcHost)
+			break
+		}
+	}
+	if volumeInfo == nil {
+		return nil, faultType, logger.LogNewErrorCodef(log, codes.Internal,
+			"failed to create volume with accessibility requirements %+v on any vCenter. Error: %+v",
+			topologyRequirement, err)
+	}
+
+	// TODO: Multi-VC: Clear up orphan volume after this point on any errors?
+	attributes := make(map[string]string)
+	attributes[common.AttributeDiskType] = common.DiskTypeBlockVolume
+	resp := &csi.CreateVolumeResponse{
+		Volume: &csi.Volume{
+			VolumeId:      volumeInfo.VolumeID.Id,
+			CapacityBytes: int64(units.FileSize(volSizeMB * common.MbInBytes)),
+			VolumeContext: attributes,
+		},
+	}
+
+	// For topology aware provisioning, populate the topology segments parameter
+	// in the CreateVolumeResponse struct.
+	var (
+		datastoreAccessibleTopology []map[string]string
+		allNodeVMs                  []*cnsvsphere.VirtualMachine
+	)
+	// Retrieve the datastoreURL of the Provisioned Volume. If CNS CreateVolume
+	// API does not return datastoreURL, retrieve this by calling QueryVolume.
+	// Otherwise, retrieve this from PlacementResults in the response of
+	// CreateVolume API.
+	datastoreURL := volumeInfo.DatastoreURL
+	if datastoreURL == "" {
+		if volumeMgr == nil {
```

Comment:
volumeMgr can't be nil here. we have set it for create volume call.

Reply:
If the CreateVolume is continued from the cached idempotency task, volumeMgr will be nil here. Just covering that case.